src/com/goide/inspections/GoAssignmentNilWithoutExplicitType.java

```diff
@@ -32,12 +35,38 @@ protected GoVisitor buildGoVisitor(@NotNull final ProblemsHolder holder, @NotNul
       @Override
       public void visitVarDeclaration(@NotNull GoVarDeclaration o) {
         for (GoVarSpec spec : o.getVarSpecList()) {
-          if (spec.getType() != null) continue;
-          for (GoExpression expr : spec.getExpressionList()) {
-            if (expr instanceof GoReferenceExpression) {
-              if (((GoReferenceExpression)expr).getIdentifier().textMatches(GoConstants.NIL)) {
-                holder.registerProblem(expr, "Cannot assign nil without explicit type", GENERIC_ERROR_OR_WARNING);
-              }
+          checkVar(spec);
+        }
+      }
+
+      @Override
+      public void visitShortVarDeclaration(@NotNull GoShortVarDeclaration o) {
+        checkVar(o);
+      }
+
+      @Override
+      public void visitConstDeclaration(@NotNull GoConstDeclaration o) {
+        for (GoConstSpec spec : o.getConstSpecList()) {
+          checkConst(spec);
+        }
+      }
+
+      private void checkVar(@NotNull GoVarSpec spec) {
+        if (spec.getType() != null) return;
+        checkExpressions(spec.getExpressionList());
+      }
+
+      private void checkConst(@NotNull GoConstSpec spec) {
+        if (spec.getType() != null) return;
+        checkExpressions(spec.getExpressionList());
+      }
+
+      private void checkExpressions(@NotNull List<GoExpression> expressions) {
+        for (GoExpression expr : expressions) {
+          if (expr instanceof GoReferenceExpressionImpl) {
+            // todo check if there is 'nil' var/const
```

Comment:
It's a weird thing. If you've declared a variable with name `nil`, say `var nil = 0` then these kinds of assignments are correct.
The following code successfully compiles.
```go
package nil
var nil = 123
var x = nil
```